programs/vote/src/vote_state/mod.rs

```diff
@@ -1770,6 +1776,76 @@ mod tests {
         );
     }
 
+    #[allow(clippy::integer_arithmetic)]
+    fn process_new_vote_state_replaced_root_vote_credits(
+        feature_set: &FeatureSet,
+        expected_credits: u64,
+    ) {
+        let mut vote_state1 = VoteState::default();
+        for i in 0..MAX_LOCKOUT_HISTORY + 2 {
+            process_slot_vote_unchecked(&mut vote_state1, i as u64);
+        }
+        assert_eq!(vote_state1.root_slot.unwrap(), 1);
+
+        // Vote credits should be 2, since two voted-on slots were "popped off the back" of the tower
+        assert_eq!(vote_state1.credits(), 2);
+
+        // Create a new vote state that represents the validator having not voted for a long time, then voting
+        // with an entirely new root that was never previously voted on.  This is valid because a vote state
+        // can include a root that it never voted on (if it votes after a very long delinquency, the new votes
+        // will have a root much newer than its most recently voted slot).
+        let mut vote_state2 = VoteState::default();
+        for i in 0..MAX_LOCKOUT_HISTORY + 2 {
+            process_slot_vote_unchecked(&mut vote_state2, (i + 10000) as u64);
+        }
+        assert_eq!(vote_state2.root_slot.unwrap(), 10001);
+
+        assert_eq!(
+            process_new_vote_state_from_votes(
+                &mut vote_state1,
+                vote_state2.votes.clone(),
+                vote_state2.root_slot,
+                None,
+                vote_state2.current_epoch(),
+                Some(feature_set)
+            ),
+            Ok(())
+        );
+
+        // The vote is valid, but no vote credits should be awarded because although there is a new root, it does not
+        // represent a slot previously voted on.
+        assert_eq!(vote_state1.credits(), expected_credits)
+    }
+
+    #[test]
+    fn test_process_new_vote_state_replaced_root_vote_credits() {
+        let mut feature_set = FeatureSet::default();
```

Comment:
`allow_votes_to_directly_update_vote_state` should be enabled as well in this test right?

Reply:
That feature isn't used anywhere in the code in question; that feature gates whether or not a VoteStateUpdate instruction would ever even make it to process_new_vote_state(), and since process_new_vote_state() is being called directly in this test, the feature is "implicity" enabled.

However, for clarity I will enable the feature in the test.